codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/DocumentClientCommandGenerator.java

```diff
@@ -0,0 +1,380 @@
+            writer.write("// Start section: $L", COMMAND_PROPERTIES_SECTION);
+            writer.pushState(COMMAND_PROPERTIES_SECTION).popState();
+            writer.write("// End section: $L", COMMAND_PROPERTIES_SECTION);
+            writer.write("");
+
+            generateCommandConstructor();
+            writer.write("");
+            generateCommandMiddlewareResolver(configType);
+
+            // Hook for adding more methods to the command.
+            writer.write("// Start section: $L", COMMAND_BODY_EXTRA_SECTION)
+                    .pushState(COMMAND_BODY_EXTRA_SECTION)
+                    .popState()
+                    .write("// End section: $L", COMMAND_BODY_EXTRA_SECTION);
+        });
+    }
+
+    private void generateCommandConstructor() {
+        writer.openBlock("constructor(readonly input: $L) {", "}", inputTypeName, () -> {
+            // The constructor can be intercepted and changed.
+            writer.write("// Start section: $L", COMMAND_CONSTRUCTOR_SECTION)
+                    .pushState(COMMAND_CONSTRUCTOR_SECTION)
+                    .write("super();")
+                    .popState()
+                    .write("// End section: $L", COMMAND_CONSTRUCTOR_SECTION);
+        });
+    }
+
+    private void generateCommandMiddlewareResolver(String configType) {
+        writer.writeDocs("@internal");
+
+        String serviceInputTypes = "ServiceInputTypes";
+        String serviceOutputTypes = "ServiceOutputTypes";
+        String handler = "Handler";
+        String middlewareStack = "MiddlewareStack";
+        writer.addImport(serviceInputTypes, serviceInputTypes, "@aws-sdk/client-dynamodb");
+        writer.addImport(serviceOutputTypes, serviceOutputTypes, "@aws-sdk/client-dynamodb");
+        writer.addImport(handler, handler, "@aws-sdk/types");
+        writer.addImport(middlewareStack, middlewareStack, "@aws-sdk/types");
+
+        writer.write("resolveMiddleware(")
+                .indent()
+                .write("clientStack: $L<$L, $L>,", middlewareStack, serviceInputTypes, serviceOutputTypes)
+                .write("configuration: $L,", configType)
+                .write("options?: $T", ApplicationProtocol.createDefaultHttpApplicationProtocol().getOptionsType())
+                .dedent();
+        writer.openBlock("): $L<$L, $L> {", "}", handler, inputTypeName, outputTypeName, () -> {
+            String marshallOptions = DocumentClientUtils.CLIENT_MARSHALL_OPTIONS;
+            String unmarshallOptions = DocumentClientUtils.CLIENT_UNMARSHALL_OPTIONS;
+            String inputKeyNodes = "inputKeyNodes";
+            String outputKeyNodes = "outputKeyNodes";
+
+            writer.write("const { $L, $L } = configuration.$L || {};", marshallOptions, unmarshallOptions,
+                    DocumentClientUtils.CLIENT_TRANSLATE_CONFIG_KEY);
+
+            if (!inputMembersWithAttr.isEmpty()) {
+                writer.openBlock("const $L = [", "];", inputKeyNodes, () -> {
```

Comment:
The input key nodes are CONSTANT values, but they are being re-declared inside the middleware handlers every time. They can be safely lifted outside these functions and declared once alongside the class.

Reply:
The same is true for the output key nodes

Reply:
The `inputKeyNodes` and `outputKeyNodes` are used only inside resolveMiddleware while calling `marshallInput` or `unmarshallOutput` respectively. Also, they are created only if they're required.

What are the benefits of moving them alongside the class?

Reply:
They are being re-declared (in memory, costs CPU) on every function invocation. Moving them alongside the class eliminates the CPU costs and reduces the memory footprint for multiple invocations.

Since they are in modules they will still only be created when used/required.

Reply:
Done in [`ffea72e` (#2097)](https://github.com/aws/aws-sdk-js-v3/pull/2097/commits/ffea72e40417981f23e5e5239e8316b462224f29)